secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,232 @@
+                    missingWordChips: newMissingWords,
+                    validationWords: newCompletionState
+                )
+            } else {
+                var newCompletion = validationWords
+                newCompletion.append(ValidationWord(groupIndex: group, word: word))
+
+                return RecoveryPhraseValidationState(
+                    phrase: phrase,
+                    missingIndices: missingIndices,
+                    missingWordChips: Array(repeating: .empty, count: RecoveryPhraseValidationState.phraseChunks),
+                    validationWords: newCompletion
+                )
+            }
+        case .complete:
+            return self
+        }
+    }
+
+    static func randomIndices() -> [Int] {
+        return (0..<phraseChunks).map { _ in
+            Int.random(in: 0 ..< wordGroupSize)
+        }
+    }
+}
+
+extension RecoveryPhrase.Group {
+    /// Returns an array of words where the word at the missing index will be an empty string
+    func words(with missingIndex: Int) -> [String] {
+        assert(missingIndex >= 0)
+        assert(missingIndex < self.words.count)
+        var wordsApplyingMissing = self.words
+        wordsApplyingMissing[missingIndex] = ""
+        return wordsApplyingMissing
+    }
+}
+
+enum RecoveryPhraseValidationAction: Equatable {
+    case updateRoute(RecoveryPhraseValidationState.Route?)
+    case reset
+    case drag(wordChip: PhraseChip.Kind, intoGroup: Int)
+    case succeed
+    case fail
+    case proceedToHome
+    case displayBackedUpPhrase
+}
+
+typealias RecoveryPhraseValidationReducer = Reducer<RecoveryPhraseValidationState, RecoveryPhraseValidationAction, BackupPhraseEnvironment>
+
+extension RecoveryPhraseValidationReducer {
+    static let `default` = RecoveryPhraseValidationReducer { state, action, environment in
+        switch action {
+        case .reset:
+            state = RecoveryPhraseValidationState.random(phrase: state.phrase)
+            return .none
+
+        case let .drag(wordChip, group):
+            state = state.apply(chip: wordChip, into: group)
+
+            // Trigger a delayed effect to proceed to the next step
+            if case .complete = state.step {
+                if state.isValid {
+                    return Effect(value: .succeed).delay(for: 1, scheduler: environment.mainQueue).eraseToEffect()
+                } else {
+                    return Effect(value: .fail).delay(for: 3, scheduler: environment.mainQueue).eraseToEffect()
+                }
+            }
+            return .none
+
+        case .succeed:
+            state.route = .success
+            return .none
+
+        case .fail:
+            state.route = .failure
+            return .none
+
+        case .updateRoute(let route):
+            state.route = route
+            return .none
+
+        case .proceedToHome:
+            return .none
+            
+        case .displayBackedUpPhrase:
+            return .none
```

Comment:
It's ok to have a `return .none` after the switch statement so you don't have to muddy up every case with the same return statement, if it works out better that way.